sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,231 @@
+                {
+                    return false;
+                }
+
+                if (IsAllowedPrimitive(elementType))
+                {
+                    return true;
+                }
+
+                return IsAllowedEnumerableValue(elementType, array);
+            }
+
+            private static bool IsAllowedListValue<T>(Type type, T value)
+            {
+                if (value == null)
+                {
+                    return true;
+                }
+
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (type.GetGenericTypeDefinition() != typeof(List<>))
+                {
+                    return false;
+                }
+
+                Type elementType = type.GetGenericArguments()[0];
+                if (IsAllowedPrimitive(elementType))
+                {
+                    return true;
+                }
+
+                return IsAllowedEnumerableValue(elementType, (IEnumerable)value);
+            }
+
+            private static bool IsAllowedDictionaryValue<T>(Type type, T value)
+            {
+                if (value == null)
+                {
+                    return true;
+                }
+
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (type.GetGenericTypeDefinition() != typeof(Dictionary<,>))
+                {
+                    return false;
+                }
+
+                Type[] types = type.GetGenericArguments();
+                if (types[0] != typeof(string))
+                {
+                    return false;
+                }
+
+                if (IsAllowedPrimitive(types[1]))
+                {
+                    return true;
+                }
+
+                return IsAllowedEnumerableValue(types[1], ((IDictionary)value).Values);
+            }
+
+            private static bool IsAllowedEnumerableValue(Type elementType, IEnumerable enumerable)
+            {
+                foreach (var item in enumerable)
+                {
+                    if (item == null)
+                    {
+                        continue;
+                    }
+
+                    // Don't allow heterogenous collections
+                    if (item.GetType() != elementType)
+                    {
+                        return false;
+                    }
+
+                    if (!IsAllowedValue(item))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            private static bool IsAllowedAnonymousValue<T>(Type type, T value, HashSet<Type> ancestorTypes)
+            {
+                if (!IsAnonymousType(type))
+                {
+                    return false;
+                }
+
+                foreach (PropertyInfo property in type.GetProperties())
+                {
```

Comment:
Can you help me understand why you prefer this to the current approach?

Reply:
I thought it will be much shorter (simpler to understand) because you will be able to use the top level IsAllowed routine. You implementation calls IsPrimitive, IsCollection, IsAnonymous in separate blocks.

Reply:
I see.  Yes, your approach is more robust with its simplicity.  I had organized it this way to avoid the "Contains()" operation lookup for cases where it wasn't needed, but if you're less concerned about the cost of that, I can reorganize it as you've suggested.

Reply:
Ah, doing what you suggested also means we need to pass the "visited" collection in at the top level.

I had originally had this factored as "Check known types; check anonymous types" to simplify this to two branches (where known types included primitives and collections).  This would move in the direction of robustness without complicating all the method signatures to take the visited collection - would you be open to this compromise?

Reply:
> I had originally had this factored as "Check known types; check anonymous types"

Keeping primitives and collections separate, though, would mean that we don't have to use reflection to obtain the property value, which might be better for performance.